src/ui/public/__tests__/events.js

```diff
@@ -0,0 +1,41 @@
+import sinon from 'sinon';
+import ngMock from 'ng_mock';
+import { EventsProvider } from 'ui/events';
+
+describe('events', function () {
+  require('test_utils/no_digest_promises').activateForSuite();
+
+  let events;
+
+  beforeEach(ngMock.module('kibana'));
+  beforeEach(ngMock.inject(function (Private) {
+    const Events = Private(EventsProvider);
+    events = new Events();
+  }));
+
+  it('calls emitted handlers asynchronously', (done) => {
+    const listenerStub = sinon.stub();
+    events.on('test', listenerStub);
+    events.emit('test');
+    sinon.assert.notCalled(listenerStub);
+
+    setTimeout(() => {
+      sinon.assert.calledOnce(listenerStub);
+      done();
+    }, 100);
```

Comment:
The `emit` should trigger on the next tick, so I think it could be a bit shorter.

Reply:
You'd think.  :(  I did as well, and initially had it at 1, but the first test broke.  I suspect it's because of the emit chain, and the initial emitChain is set to Promise.resolve, so maybe it actually takes two cycles, not one?  I didn't bother digging deeper though and just set it to 100.

btw, I agree this async emit stuff is whacky.  Actually like a year ago I had a long chat with @spalger about how to remove the reliance on the angular Promise class in here, in an effort to split out framework agnostic business portions (and a lot of the code I wanted to convert over ended up needing the custom Promise class just because of this event emitter).

The conversation was long ago but if I recall correctly, the whole reason for this complication is to automatically insert digest loops with the use of angular promises, so the UI will automatically update after every handler is called.  We did discuss an alternative, which I never got around to implementing, but I think it still involved using Promises to inject digest loops.

Given the recent conversation we had about angular vs native promises (https://github.com/elastic/kibana/issues/13855) I think the best way forward would be to revert back to a normal event emitter and get rid of all this confusing promise wrapping, and require angular callbacks to use angular promises for their digest loops, rather than relying on this very deeply nested logic.

Given how deeply entrenched state management/event emitter is in our angular code, it may be risky (though IMO, still worth investigating).